### PR TITLE
DRILL-4474: Ensure that ConvertCountToDirectScan only pushes through project when project is trivial.

### PR DESCRIPTION
**Pull request by jacques-n**
There are probably additional optimizations here but as the code stood, we completely ignored whatever was in the project above the scan below the count(x).

### CODE REVIEW

**Comment by amansinha100** [comment]
+1.   Yes, seems like the project was not being checked at all and the nullability check did not work for expressions inside the count aggregate.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/ConvertCountToDirectScan.java [hunk 1]
```diff
@@ -103,6 +104,10 @@ public void onMatch(RelOptRuleCall call) {
       return;
     }
 
+    if (proj != null && !ProjectRemoveRule.isTrivial(proj)) {
```
Comment (jinfengni) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/ConvertCountToDirectScan.java]: I feel that this check might over-kill some optimization opportunity.  For example,

select count(100) 
from `parquetTable`;

In this case, count(100) is equal to rowcount in parquet table. However, the project is not a trial project, meaning the new code will disable the optimization.

Comment (jinfengni) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/ConvertCountToDirectScan.java]: We have a check whether the input to count() is nullable in Line 115. In theory, if the input is non-nullable, then count(non-nullalbe expression) = rowcount.

My guess is that the query (case expression) with incorrect result is caused by the wrong type resolution for the case expression.

Comment (jinfengni) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/ConvertCountToDirectScan.java]: With the patch, the following query will not use directScan. 

{code}
select count(*) from cp.`tpch/nation.parquet`;
{code}

{code}
00-00    Screen : rowType = RecordType(BIGINT EXPR$0): rowcount = 1.0, cumulative cost = {75.1 rows, 425.1 cpu, 0.0 io, 0.0 network, 0.0 memory}, id = 169
00-01      Project(EXPR$0=[$0]) : rowType = RecordType(BIGINT EXPR$0): rowcount = 1.0, cumulative cost = {75.0 rows, 425.0 cpu, 0.0 io, 0.0 network, 0.0 memory}, id = 168
00-02        StreamAgg(group=[{}], EXPR$0=[COUNT()]) : rowType = RecordType(BIGINT EXPR$0): rowcount = 1.0, cumulative cost = {75.0 rows, 425.0 cpu, 0.0 io, 0.0 network, 0.0 memory}, id = 167
00-03          Project($f0=[0]) : rowType = RecordType(INTEGER $f0): rowcount = 25.0, cumulative cost = {50.0 rows, 125.0 cpu, 0.0 io, 0.0 network, 0.0 memory}, id = 166
00-04            Scan(groupscan=[ParquetGroupScan [entries=[ReadEntryWithPath [path=classpath:/tpch/nation.parquet]], selectionRoot=classpath:/tpch/nation.parquet, numFiles=1, usedMetadataFile=false, columns=[]]]) : rowType = RecordType(): rowcount = 25.0, cumulative cost = {25.0 rows, 25.0 cpu, 0.0 io, 0.0 network, 0.0 memory}, id = 165
{code}

I debug a bit. Seems Line 115 is fine. But something is worng in the code Line 117 - 123.

**Comment by amansinha100** [comment]
Agree with @jinfengni that the current fix can cause performance regression for simpler count queries.  I will change my review to -1 and let's see how to get the proper nullability check.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/ConvertCountToDirectScan.java (hunk 1 above)
Comment (jinfengni) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/ConvertCountToDirectScan.java]: I have a patch, which works fine for Jacque's new unit test. It continues to use directScan for simple count query. The patch is pending pre-commit & unit test run. Will update results shortly.

**Pull request closed without merging**